core/auth/src/main/java/software/amazon/awssdk/auth/signer/internal/BaseEventStreamAsyncAws4Signer.java

```diff
@@ -259,4 +267,35 @@ public void subscribe(Subscriber<? super ByteBuffer> s) {
             return transformedRequestBody.contentLength();
         }
     }
+
+    private static String toDebugString(Message m, boolean truncatePayload) {
+        StringBuilder sb = new StringBuilder("Message = {headers={");
+        Map<String, HeaderValue> headers = m.getHeaders();
+
+        Iterator<Map.Entry<String, HeaderValue>> headersIter = headers.entrySet().iterator();
+
+        while (headersIter.hasNext()) {
+            Map.Entry<String, HeaderValue> h = headersIter.next();
+
+            sb.append(h.getKey()).append("={").append(h.getValue().toString()).append("}");
+
+            if (headersIter.hasNext()) {
+                sb.append(", ");
+            }
+        }
+
+        sb.append("}, payload=");
+
+        byte[] payload = m.getPayload();
+
+        byte[] payloadToLog;
+        if (truncatePayload) {
+            // Would be nice if BinaryUtils.toHex() could take an array index range instead so we don't need to copy
+            payloadToLog = Arrays.copyOf(payload, Math.min(32, payload.length));
```

Comment:
Could payload.length be 0? what happens then?

Reply:
This should be fine, it will generate an empty string